### PR TITLE
Add Docker container environment variables as tags. Only whitelisted #2580

### PR DESCRIPTION
**Pull request by rsingh2411**
Add Docker container environment variables as tags. Only whitelisted #2580
PR for the above issue.

### Required for all PRs:

- [x] CHANGELOG.md updated (we recommend not updating this until the PR has been approved by a maintainer)
- [x] Sign [CLA](https://influxdata.com/community/cla/) (if not already signed)

### CODE REVIEW

File: plugins/inputs/docker/docker.go [hunk 1]
```diff
@@ -296,6 +311,22 @@ func (d *Docker) gatherContainer(
...
+	// Add whitelisted environment variables to tags
+	if len(d.GatherEnvironment) > 0 {
+		info, err := inspectWrapper(d.client, ctx, container.ID)
+		if err != nil {
+			return fmt.Errorf("Error getting docker container inspect: %s", err.Error())
+		}
+
+		for _, envvar := range info.Config.Env {
+			kvs := strings.SplitN(envvar, "=", 2)
+			if !sliceContains(kvs[0], d.GatherEnvironment) {
```
Comment (danielnelson) [inline plugins/inputs/docker/docker.go]: Don't use this function since it is private to the libraries docker package, its just a coincidence that we are also in a docker package.  I think this loop would be better if we ranged over d.GatherEnvironment and split only if the prefix matches.

File: plugins/inputs/docker/docker.go [hunk 2]
```diff
@@ -296,6 +311,22 @@ func (d *Docker) gatherContainer(
...
+				continue
+			}
+			tags[kvs[0]] = kvs[1]
```
Comment (danielnelson) [inline plugins/inputs/docker/docker.go]: Lets be extra careful and check the len of kvs, since SplitN could return a slice of length 1.

File: plugins/inputs/docker/docker_test.go [hunk 3]
```diff
@@ -247,8 +247,9 @@ func testStats() *types.StatsJSON {
 func TestDockerGatherInfo(t *testing.T) {
 	var acc testutil.Accumulator
 	d := Docker{
-		client:  nil,
-		testing: true,
+		client:            nil,
+		testing:           true,
+		GatherEnvironment: []string{"ENVVAR1", "ENVVAR2"},
```
Comment (danielnelson) [inline plugins/inputs/docker/docker_test.go]: Can you add tests for envvars that are not found in the container, envvars with zero length values, multiple '=' signs, and any other edge cases you can think of.

**Comment by danielnelson** [comment]
Please update the docker README.md too.

File: plugins/inputs/docker/docker.go [hunk 4]
```diff
@@ -98,6 +111,8 @@ var sampleConfig = `
   perdevice = true
   ## Whether to report for each container total blkio and network stats or not
   total = false
+  ## Which environment variables should we use as a tag
+  gather_environment = ["ENV", "DC"]
```
Comment (rsingh2411) [inline plugins/inputs/docker/docker.go]: Done

File: plugins/inputs/docker/docker.go [hunk 5]
```diff
@@ -296,6 +311,22 @@ func (d *Docker) gatherContainer(
 		tags[k] = label
 	}
 
+	// Add whitelisted environment variables to tags
+	if len(d.GatherEnvironment) > 0 {
+		info, err := inspectWrapper(d.client, ctx, container.ID)
+		if err != nil {
+			return fmt.Errorf("Error getting docker container inspect: %s", err.Error())
```
Comment (rsingh2411) [inline plugins/inputs/docker/docker.go]: Done

File: plugins/inputs/docker/docker.go (hunk 1 above)
Comment (rsingh2411) [inline plugins/inputs/docker/docker.go]: Done

File: plugins/inputs/docker/docker.go (hunk 2 above)
Comment (rsingh2411) [inline plugins/inputs/docker/docker.go]: Done

File: plugins/inputs/docker/docker_test.go (hunk 3 above)
Comment (rsingh2411) [inline plugins/inputs/docker/docker_test.go]: I have added mentioned negative edge cases.

**Comment by rsingh2411** [comment]
Review changes are incorporated. Please have a look

**Comment by rsingh2411** [comment]
There were some conflicts due to Docker: optionally add labels as tags (#2425) , these changes,
apologies for redundant commits. Tests are passing successfully

File: plugins/inputs/docker/README.md [hunk 6]
```diff
@@ -30,12 +30,13 @@ for the stat structure can be found
   perdevice = true
   ## Whether to report for each container total blkio and network stats or not
   total = false
+  ## Which environment variables should we use as a tag
+  gather_environment = ["JAVA_HOME", "HEAP_SIZE"]
```
Comment (danielnelson) [inline plugins/inputs/docker/README.md]: Maybe we should rename this option to `tag_env`?

File: plugins/inputs/docker/docker.go [hunk 7]
```diff
@@ -322,6 +337,27 @@ func (d *Docker) gatherContainer(
...
 
+	// Add whitelisted environment variables to tags
+	if len(d.GatherEnvironment) > 0 {
+		info, err := inspectWrapper(d.client, ctx, container.ID)
+		if err != nil {
+			return fmt.Errorf("Error inspecting docker container: %s", err.Error())
+		}
+		for _, envvar := range info.Config.Env {
+			for _, prefix := range d.GatherEnvironment {
+				if strings.Count(envvar, "=") > 1 {
```
Comment (danielnelson) [inline plugins/inputs/docker/docker.go]: I think we should't skip environment variables like: `FOO=BAR=BAZ`, also add a testcase for this.

File: plugins/inputs/docker/docker.go [hunk 8]
```diff
@@ -322,6 +337,27 @@ func (d *Docker) gatherContainer(
...
+					continue
+				}
+				if strings.HasPrefix(envvar, prefix) {
```
Comment (danielnelson) [inline plugins/inputs/docker/docker.go]: I think items with a common prefix will get picked up.  Try adding a test case for if there are environment variables `FOO` and `FOOBAR`, and specify `FOO` in the config.

File: plugins/inputs/docker/README.md (hunk 6 above)
Comment (rsingh2411) [inline plugins/inputs/docker/README.md]: Done

File: plugins/inputs/docker/docker.go (hunk 7 above)
Comment (rsingh2411) [inline plugins/inputs/docker/docker.go]: Done, added the testcase

File: plugins/inputs/docker/docker.go (hunk 8 above)
Comment (rsingh2411) [inline plugins/inputs/docker/docker.go]: Added an testcase for this as well

**Comment by rsingh2411** [comment]
I have incorporated review changes. Please have a look

File: plugins/inputs/docker/docker_test.go [hunk 9]
```diff
@@ -298,8 +298,9 @@ func TestDockerGatherLabels(t *testing.T) {
 func TestDockerGatherInfo(t *testing.T) {
 	var acc testutil.Accumulator
 	d := Docker{
-		client:  nil,
-		testing: true,
+		client:         nil,
+		testing:        true,
+		TagEnvironment: []string{"ENVVAR1", "ENVVAR2", "ENVVAR3", "ENVVAR4", "ENVVAR5", "ENVVAR6", "ENVVAR7", "ENVVAR8", "ENVVAR9"},
```
Comment (danielnelson) [inline plugins/inputs/docker/docker_test.go]: Please line wrap this after each element, not a hard rule but try to stick to about 80 chars per line.

File: plugins/inputs/docker/docker_test.go [hunk 10]
```diff
@@ -345,6 +346,12 @@ func TestDockerGatherInfo(t *testing.T) {
 			"cpu":               "cpu3",
 			"container_version": "v2.2.2",
 			"engine_host":       "absol",
+			"ENVVAR1":           "loremipsum",
+			"ENVVAR2":           "dolorsitamet",
+			"ENVVAR3":           "ubuntu:10.04",
+			"ENVVAR6":           " ",
```
Comment (danielnelson) [inline plugins/inputs/docker/docker_test.go]: Even though this is technically correct, I think we shouldn't create a field if it contains only whitespace or an empty string.

File: plugins/inputs/docker/docker_test.go [hunk 11]
```diff
@@ -345,6 +346,12 @@ func TestDockerGatherInfo(t *testing.T) {
...
+			"ENVVAR7":           "ENVVAR9",
```
Comment (danielnelson) [inline plugins/inputs/docker/docker_test.go]: Value should be `ENVVAR8=ENVVAR9`

File: plugins/inputs/docker/docker_test.go [hunk 12]
```diff
@@ -345,6 +346,12 @@ func TestDockerGatherInfo(t *testing.T) {
...
+			"ENVVAR8":           "ENVVAR9",
```
Comment (danielnelson) [inline plugins/inputs/docker/docker_test.go]: There shouldn't be an `ENVVAR8`

File: plugins/inputs/docker/docker_test.go (hunk 9 above)
Comment (rsingh2411) [inline plugins/inputs/docker/docker_test.go]: Will be done.

File: plugins/inputs/docker/docker_test.go (hunk 11 above)
Comment (rsingh2411) [inline plugins/inputs/docker/docker_test.go]: Apologies, didn't get this comment
say ENVVAR7=ENVVAR8=ENVVAR9
then tags would be
ENVVAR7=ENVVAR9
ENVVAR8=ENVVAR9
please correct me if my understanding is wrong.

Comment (danielnelson) [inline plugins/inputs/docker/docker_test.go]: The tagkey should be `ENVVAR7` and the tagvalue should be `ENVVAR8=ENVVAR9`

File: plugins/inputs/docker/docker_test.go (hunk 9 above)
Comment (rsingh2411) [inline plugins/inputs/docker/docker_test.go]: Done

File: plugins/inputs/docker/docker_test.go (hunk 11 above)
Comment (rsingh2411) [inline plugins/inputs/docker/docker_test.go]: Done

File: plugins/inputs/docker/docker_test.go (hunk 12 above)
Comment (rsingh2411) [inline plugins/inputs/docker/docker_test.go]: Done

**Comment by danielnelson** [comment]
@rsingh2411 Can you add an entry to the 1.4 feature section of the CHANGELOG?

@nhaugo Did you want to review?
